develop/tutorials/articles/193-adaptive-media/00-using-adaptive-media-apis-intro.markdown

```diff
@@ -0,0 +1,283 @@
+                                    AdaptiveMediaAttributeConverterUtil::parseInt,
+                                    ImageAdaptiveMediaAttribute::_intDistance);
+    
+            public static final AdaptiveMediaAttribute
+                    <ImageAdaptiveMediaProcessor, Integer> IMAGE_WIDTH =
+                            new AdaptiveMediaAttribute<>(
+                                    "width", 
+                                    AdaptiveMediaAttributeConverterUtil::parseInt,
+                                    ImageAdaptiveMediaAttribute::_intDistance);
+    
+            /**
+             * Returns a string-attribute map containing the available
+             * name-attribute pairs.
+             *
+             * @return the list of available attributes
+             */
+            public static Map<String, AdaptiveMediaAttribute<?, ?>>
+                    allowedAttributes() {
+    
+                    return _allowedAttributes;
+            }
+    
+            private static int _intDistance(int i1, int i2) {
+                    return i1 - i2;
+            }
+    
+            private ImageAdaptiveMediaAttribute() {
+            }
+    
+            private static final Map<String, AdaptiveMediaAttribute<?, ?>>
+                    _allowedAttributes = new HashMap<>();
+    
+            static {
+                    _allowedAttributes.put(
+                            ImageAdaptiveMediaAttribute.IMAGE_WIDTH.getName(),
+                            ImageAdaptiveMediaAttribute.IMAGE_WIDTH);
+                    _allowedAttributes.put(
+                            ImageAdaptiveMediaAttribute.IMAGE_HEIGHT.getName(),
+                            ImageAdaptiveMediaAttribute.IMAGE_HEIGHT);
+            }
+    
+    }
+
+First, the attributes are individually added to a new `AdaptiveMediaAttribute`
+object. The `AdaptiveMediaAttribute` class takes a processor type and
+attribute type as parameters. The example above uses 
+`ImageAdaptiveMediaProcessor` and an `Integer` for the attribute type. A new 
+instance of the `AdaptiveMediaAttribute` is then created passing the 
+`String name`, `Function<String, V> converter`, and a `Comparator<V> comparator` 
+arguments. For example, the `height` attribute above uses `height` as the name, 
+`AdaptiveMediaAttributeConverterUtil::parseInt` as the converter function, and 
+`ImageAdaptiveMediaAttribute::_intDistance` as the comparator.
+Finally, the new `height` and `width` attributes for the 
+`ImageAdaptiveMediaAttribute` are added to the `_allowedAttributes` HashMap.
+
+The contents of a media are exposed via an InputStream. Once requested
+with the [`getInputStream()` method](https://github.com/liferay/com-liferay-adaptive-media/blob/6356d57a76050764d52a8dccce07578cc53b72b1/adaptive-media-api/src/main/java/com/liferay/adaptive/media/AdaptiveMedia.java#L45-L52), 
+consumers are responsible for closing it when finished.
+
+<!-- Is there an example of using and closing an InputStream? Is this 
```

Comment:
Is there an example of using and closing an InputStream? 

Is this ImageProcessor class an example of one? [https://github.com/liferay/com-liferay-adaptive-media/blob/5220d012bc037a1736746eddc9a0ab9aa90ea8d4/adaptive-media-image-impl/src/main/java/com/liferay/adaptive/media/image/internal/util/ImageProcessor.java](https://github.com/liferay/com-liferay-adaptive-media/blob/5220d012bc037a1736746eddc9a0ab9aa90ea8d4/adaptive-media-image-impl/src/main/java/com/liferay/adaptive/media/image/internal/util/ImageProcessor.java)-->